audit_log/admin.py

```diff
@@ -0,0 +1,228 @@
+        return (
+            super()
+            .get_queryset(request)
+            .with_audit_log_and_request(
+                request=request,
+                operation=Operation.READ.value,
+            )
+        )
+
+    def save_model(self, request, obj, form, change):
+        """
+        Given a model instance save it to the database.
+        """
+        # TODO: add a change log to the audit log message
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.UPDATE.value if change else Operation.CREATE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().save_model(request, obj, form, change)
+
+    def delete_model(self, request, obj):
+        """
+        Given a model instance delete it from the database.
+        """
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.DELETE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().delete_model(request, obj)
+
+    def delete_queryset(self, request, queryset):
+        """Given a queryset, delete it from the database."""
+        super().delete_queryset(
+            request,
+            queryset.with_audit_log_and_request(
+                request=request, operation=Operation.DELETE.value
+            ),
+        )
+
+
+class AuditLogMessageOperationListFilter(admin.SimpleListFilter):
+    # Human-readable title which will be displayed in the
+    # right admin sidebar just above the filter options.
+    title = _("operation")
+
+    # Parameter for the filter that will be used in the URL query.
+    parameter_name = "operation"
+
+    options = [operation.value for operation in Operation]
+
+    def lookups(self, request, model_admin):
+        return map(lambda o: (o, o.upper()), self.options)
+
+    def queryset(self, request, queryset):
+        if self.value() in self.options:
+            return queryset.filter(message__icontains=self.value())
+        return queryset
+
+
+class AuditLogMessageRoleListFilter(admin.SimpleListFilter):
+    # Human-readable title which will be displayed in the
+    # right admin sidebar just above the filter options.
+    title = _("role")
+
+    # Parameter for the filter that will be used in the URL query.
+    parameter_name = "role"
+
+    options = [role.value for role in Role]
+
+    def lookups(self, request, model_admin):
+        return map(lambda o: (o, o.upper()), self.options)
+
+    def queryset(self, request, queryset):
+        if self.value() in self.options:
+            return queryset.filter(message__icontains=self.value())
+        return queryset
+
+
+class LargeTablePaginator(Paginator):
```

Comment:
```suggestion
class FastImpreciseTablePaginator(Paginator):
```
That might be more accurate description of this paginator class. And also such that if one sees it being used then they get an accurate description of what's going on, they won't expect it to give precise paginations i.e. this would add transparency/clarity.

Reply:
Used this new name